File: booksamples/appendix-01/src/main/java/org/vulkanb/eng/graph/vk/PhysicalDevice.java
package org.vulkanb.eng.graph.vk;

import org.lwjgl.PointerBuffer;
import org.lwjgl.system.MemoryStack;
import org.lwjgl.vulkan.*;
import org.tinylog.Logger;

import java.nio.IntBuffer;
import java.util.*;

import static org.lwjgl.vulkan.VK11.*;
import static org.vulkanb.eng.graph.vk.VulkanUtils.vkCheck;

public class PhysicalDevice {

    private final VkExtensionProperties.Buffer vkDeviceExtensions;
    private final VkPhysicalDeviceMemoryProperties vkMemoryProperties;
    private final VkPhysicalDevice vkPhysicalDevice;
    private final VkPhysicalDeviceFeatures vkPhysicalDeviceFeatures;
    private final VkPhysicalDeviceProperties vkPhysicalDeviceProperties;
    private final VkQueueFamilyProperties.Buffer vkQueueFamilyProps;
    private CheckPointExtension checkPointExtension;
    private Set<Integer> suportedSampleCount;

    private PhysicalDevice(VkPhysicalDevice vkPhysicalDevice) {
        try (MemoryStack stack = MemoryStack.stackPush()) {
            this.vkPhysicalDevice = vkPhysicalDevice;

            IntBuffer intBuffer = stack.mallocInt(1);

            // Get device properties
            vkPhysicalDeviceProperties = VkPhysicalDeviceProperties.calloc();
            vkGetPhysicalDeviceProperties(vkPhysicalDevice, vkPhysicalDeviceProperties);

            // Get device extensions
            vkCheck(vkEnumerateDeviceExtensionProperties(vkPhysicalDevice, (String) null, intBuffer, null),
                    "Failed to get number of device extension properties");
            vkDeviceExtensions = VkExtensionProperties.calloc(intBuffer.get(0));
            vkCheck(vkEnumerateDeviceExtensionProperties(vkPhysicalDevice, (String) null, intBuffer, vkDeviceExtensions),
                    "Failed to get extension properties");

            checkPointExtension = calCheckPointExtension(vkDeviceExtensions);
            suportedSampleCount = calSupportedSampleCount(vkPhysicalDeviceProperties);

            // Get Queue family properties
            vkGetPhysicalDeviceQueueFamilyProperties(vkPhysicalDevice, intBuffer, null);
            vkQueueFamilyProps = VkQueueFamilyProperties.calloc(intBuffer.get(0));
            vkGetPhysicalDeviceQueueFamilyProperties(vkPhysicalDevice, intBuffer, vkQueueFamilyProps);

            vkPhysicalDeviceFeatures = VkPhysicalDeviceFeatures.calloc();
            vkGetPhysicalDeviceFeatures(vkPhysicalDevice, vkPhysicalDeviceFeatures);

            // Get Memory information and properties
            vkMemoryProperties = VkPhysicalDeviceMemoryProperties.calloc();
            vkGetPhysicalDeviceMemoryProperties(vkPhysicalDevice, vkMemoryProperties);
        }
    }

    public static PhysicalDevice createPhysicalDevice(Instance instance, String prefferredDeviceName) {
        Logger.debug("Selecting physical devices");
        PhysicalDevice selectedPhysicalDevice = null;
        try (MemoryStack stack = MemoryStack.stackPush()) {
            // Get available devices
            PointerBuffer pPhysicalDevices = getPhysicalDevices(instance, stack);
            int numDevices = pPhysicalDevices != null ? pPhysicalDevices.capacity() : 0;
            if (numDevices <= 0) {
                throw new RuntimeException("No physical devices found");
            }

            // Populate available devices
            List<PhysicalDevice> devices = new ArrayList<>();
            for (int i = 0; i < numDevices; i++) {
                VkPhysicalDevice vkPhysicalDevice = new VkPhysicalDevice(pPhysicalDevices.get(i), instance.getVkInstance());
                PhysicalDevice physicalDevice = new PhysicalDevice(vkPhysicalDevice);

                String deviceName = physicalDevice.getDeviceName();
                if (physicalDevice.hasGraphicsQueueFamily() && physicalDevice.hasKHRSwapChainExtension()) {
                    Logger.debug("Device [{}] supports required extensions", deviceName);
                    if (prefferredDeviceName != null && prefferredDeviceName.equals(deviceName)) {
                        selectedPhysicalDevice = physicalDevice;
                        break;
                    }
                    devices.add(physicalDevice);
                } else {
                    Logger.debug("Device [{}] does not support required extensions", deviceName);
                    physicalDevice.cleanup();
                }
            }

            // No preferred device or it does not meet requirements, just pick the first one
            selectedPhysicalDevice = selectedPhysicalDevice == null && !devices.isEmpty() ? devices.remove(0) : selectedPhysicalDevice;

            // Clean up non-selected devices
            for (PhysicalDevice physicalDevice : devices) {
                physicalDevice.cleanup();
            }

            if (selectedPhysicalDevice == null) {
                throw new RuntimeException("No suitable physical devices found");
            }
            Logger.debug("Selected device: [{}]", selectedPhysicalDevice.getDeviceName());
        }

        return selectedPhysicalDevice;
    }

    protected static PointerBuffer getPhysicalDevices(Instance instance, MemoryStack stack) {
        PointerBuffer pPhysicalDevices;
        // Get number of physical devices
        IntBuffer intBuffer = stack.mallocInt(1);
        vkCheck(vkEnumeratePhysicalDevices(instance.getVkInstance(), intBuffer, null),
                "Failed to get number of physical devices");
        int numDevices = intBuffer.get(0);
        Logger.debug("Detected {} physical device(s)", numDevices);

        // Populate physical devices list pointer
        pPhysicalDevices = stack.mallocPointer(numDevices);
        vkCheck(vkEnumeratePhysicalDevices(instance.getVkInstance(), intBuffer, pPhysicalDevices),
                "Failed to get physical devices");
        return pPhysicalDevices;
    }

    private CheckPointExtension calCheckPointExtension(VkExtensionProperties.Buffer vkDeviceExtensions) {
        CheckPointExtension result = CheckPointExtension.NONE;

        int numExtensions = vkDeviceExtensions != null ? vkDeviceExtensions.capacity() : 0;
        for (int i = 0; i < numExtensions; i++) {
            String extensionName = vkDeviceExtensions.get(i).extensionNameString();
            if (NVDeviceDiagnosticCheckpoints.VK_NV_DEVICE_DIAGNOSTIC_CHECKPOINTS_EXTENSION_NAME.equals(extensionName)) {
                result = CheckPointExtension.NVIDIA;
                break;
            } else if (AMDBufferMarker.VK_AMD_BUFFER_MARKER_EXTENSION_NAME.equals(extensionName)) {
                result = CheckPointExtension.AMD;
                break;
            }
        }
        return result;
    }

    private Set<Integer> calSupportedSampleCount(VkPhysicalDeviceProperties devProps) {
        Set<Integer> result = new HashSet<>();
        long colorCounts = Integer.toUnsignedLong(vkPhysicalDeviceProperties.limits().framebufferColorSampleCounts());
        Logger.debug("Color max samples: {}", colorCounts);
        long depthCounts = Integer.toUnsignedLong(devProps.limits().framebufferDepthSampleCounts());
        Logger.debug("Depth max samples: {}", depthCounts);
        int counts = (int) (Math.min(colorCounts, depthCounts));
        Logger.debug("Max samples: {}", depthCounts);

        result.add(VK_SAMPLE_COUNT_1_BIT);
        if ((counts & VK_SAMPLE_COUNT_64_BIT) > 0) {
            result.add(VK_SAMPLE_COUNT_64_BIT);
        }
        if ((counts & VK_SAMPLE_COUNT_32_BIT) > 0) {
            result.add(VK_SAMPLE_COUNT_32_BIT);
        }
        if ((counts & VK_SAMPLE_COUNT_16_BIT) > 0) {
            result.add(VK_SAMPLE_COUNT_16_BIT);
        }
        if ((counts & VK_SAMPLE_COUNT_8_BIT) > 0) {
            result.add(VK_SAMPLE_COUNT_8_BIT);
        }
        if ((counts & VK_SAMPLE_COUNT_4_BIT) > 0) {
            result.add(VK_SAMPLE_COUNT_4_BIT);
        }
        if ((counts & VK_SAMPLE_COUNT_2_BIT) != 0) {
            result.add(VK_SAMPLE_COUNT_2_BIT);
        }

        return result;
    }

    public void cleanup() {
        if (Logger.isDebugEnabled()) {
            Logger.debug("Destroying physical device [{}]", vkPhysicalDeviceProperties.deviceNameString());
        }
        vkMemoryProperties.free();
        vkPhysicalDeviceFeatures.free();
        vkQueueFamilyProps.free();
        vkDeviceExtensions.free();
        vkPhysicalDeviceProperties.free();
    }

    public CheckPointExtension getCheckPointExtension() {
        return checkPointExtension;
    }

    public String getDeviceName() {
        return vkPhysicalDeviceProperties.deviceNameString();
    }

    public VkPhysicalDeviceMemoryProperties getVkMemoryProperties() {
        return vkMemoryProperties;
    }

    public VkPhysicalDevice getVkPhysicalDevice() {
        return vkPhysicalDevice;
    }

    public VkPhysicalDeviceFeatures getVkPhysicalDeviceFeatures() {
        return vkPhysicalDeviceFeatures;
    }

    public VkPhysicalDeviceProperties getVkPhysicalDeviceProperties() {
        return vkPhysicalDeviceProperties;
    }

    public VkQueueFamilyProperties.Buffer getVkQueueFamilyProps() {
        return vkQueueFamilyProps;
    }

    private boolean hasGraphicsQueueFamily() {
        boolean result = false;
        int numQueueFamilies = vkQueueFamilyProps != null ? vkQueueFamilyProps.capacity() : 0;
        for (int i = 0; i < numQueueFamilies; i++) {
            VkQueueFamilyProperties familyProps = vkQueueFamilyProps.get(i);
            if ((familyProps.queueFlags() & VK_QUEUE_GRAPHICS_BIT) != 0) {
                result = true;
                break;
            }
        }
        return result;
    }

    private boolean hasKHRSwapChainExtension() {
        boolean result = false;
        int numExtensions = vkDeviceExtensions != null ? vkDeviceExtensions.capacity() : 0;
        for (int i = 0; i < numExtensions; i++) {
            String extensionName = vkDeviceExtensions.get(i).extensionNameString();
            if (KHRSwapchain.VK_KHR_SWAPCHAIN_EXTENSION_NAME.equals(extensionName)) {
                result = true;
                break;
            }
        }
        return result;
    }

    public boolean supportsSampleCount(int numSamples) {
        return suportedSampleCount.contains(numSamples);
    }

    public enum CheckPointExtension {
        NONE, NVIDIA, AMD;
    }
}

File: booksamples/chapter-05/kotlin/src/main/kotlin/org/vulkanb/eng/graph/vk/PhysicalDevice.kt
package org.vulkanb.eng.graph.vk

import io.github.oshai.kotlinlogging.KotlinLogging
import org.lwjgl.PointerBuffer
import org.lwjgl.system.MemoryStack
import org.lwjgl.vulkan.*
import org.lwjgl.vulkan.VK11.*
import org.vulkanb.eng.graph.vk.VulkanUtils.vkCheck

private val logger = KotlinLogging.logger {}

open class PhysicalDevice private constructor(val vkPhysicalDevice: VkPhysicalDevice) {
    val vkDeviceExtensions: VkExtensionProperties.Buffer
    val vkMemoryProperties: VkPhysicalDeviceMemoryProperties
    val vkPhysicalDeviceFeatures: VkPhysicalDeviceFeatures
    val vkPhysicalDeviceProperties: VkPhysicalDeviceProperties
    val vkQueueFamilyProps: VkQueueFamilyProperties.Buffer

    init {
        MemoryStack.stackPush().use { stack ->
            val intBuffer = stack.mallocInt(1)

            // Get device properties
            vkPhysicalDeviceProperties = VkPhysicalDeviceProperties.calloc()
            vkGetPhysicalDeviceProperties(vkPhysicalDevice, vkPhysicalDeviceProperties)

            // Get device extensions
            vkEnumerateDeviceExtensionProperties(vkPhysicalDevice, null as String?, intBuffer, null)
                .vkCheck("Failed to get number of device extension properties")

            vkDeviceExtensions = VkExtensionProperties.calloc(intBuffer.get(0))
            vkEnumerateDeviceExtensionProperties(vkPhysicalDevice, null as String?, intBuffer, vkDeviceExtensions)
                .vkCheck("Failed to get extension properties")

            // Get Queue family properties
            vkGetPhysicalDeviceQueueFamilyProperties(vkPhysicalDevice, intBuffer, null)
            vkQueueFamilyProps = VkQueueFamilyProperties.calloc(intBuffer.get(0))
            vkGetPhysicalDeviceQueueFamilyProperties(vkPhysicalDevice, intBuffer, vkQueueFamilyProps)
            vkPhysicalDeviceFeatures = VkPhysicalDeviceFeatures.calloc()
            vkGetPhysicalDeviceFeatures(vkPhysicalDevice, vkPhysicalDeviceFeatures)

            // Get Memory information and properties
            vkMemoryProperties = VkPhysicalDeviceMemoryProperties.calloc()
            vkGetPhysicalDeviceMemoryProperties(vkPhysicalDevice, vkMemoryProperties)
        }
    }

    fun cleanup() {
        logger.debug { "Destroying physical device [${vkPhysicalDeviceProperties.deviceNameString()}]" }
        vkMemoryProperties.free()
        vkPhysicalDeviceFeatures.free()
        vkQueueFamilyProps.free()
        vkDeviceExtensions.free()
        vkPhysicalDeviceProperties.free()
    }

    val deviceName: String
        get() = vkPhysicalDeviceProperties.deviceNameString()

    private fun hasGraphicsQueueFamily(): Boolean {
        var result = false
        val numQueueFamilies = vkQueueFamilyProps.capacity()
        for (i in 0..<numQueueFamilies) {
            val familyProps = vkQueueFamilyProps.get(i)
            if (familyProps.queueFlags() and VK_QUEUE_GRAPHICS_BIT != 0) {
                result = true
                break
            }
        }
        return result
    }

    private fun hasKHRSwapChainExtension(): Boolean {
        var result = false
        val numExtensions = vkDeviceExtensions.capacity()
        for (i in 0..<numExtensions) {
            val extensionName = vkDeviceExtensions.get(i).extensionNameString()
            if (KHRSwapchain.VK_KHR_SWAPCHAIN_EXTENSION_NAME == extensionName) {
                result = true
                break
            }
        }
        return result
    }

    companion object {
        fun createPhysicalDevice(preferredDeviceName: String?): PhysicalDevice {
            logger.debug { "Selecting physical devices" }
            var selectedPhysicalDevice: PhysicalDevice? = null
            MemoryStack.stackPush().use { stack ->
                // Get available devices
                val pPhysicalDevices = getPhysicalDevices(stack)
                val numDevices = pPhysicalDevices.capacity()
                if (numDevices <= 0) {
                    throw RuntimeException("No physical devices found")
                }

                // Populate available devices
                val devices: MutableList<PhysicalDevice> = ArrayList()
                for (i in 0..<numDevices) {
                    val vkPhysicalDevice = VkPhysicalDevice(pPhysicalDevices.get(i), Instance.vkInstance)
                    val physicalDevice = PhysicalDevice(vkPhysicalDevice)
                    val deviceName = physicalDevice.deviceName
                    if (physicalDevice.hasGraphicsQueueFamily() && physicalDevice.hasKHRSwapChainExtension()) {
                        logger.debug { "Device [$deviceName] supports required extensions" }
                        if (preferredDeviceName != null && preferredDeviceName == deviceName) {
                            selectedPhysicalDevice = physicalDevice
                            break
                        }
                        devices.add(physicalDevice)
                    } else {
                        logger.debug { "Device [$deviceName] does not support required extensions" }
                        physicalDevice.cleanup()
                    }
                }

                // No preferred device or it does not meet requirements, just pick the first one
                selectedPhysicalDevice =
                    if (selectedPhysicalDevice == null && devices.isNotEmpty()) devices.removeAt(0) else selectedPhysicalDevice

                // Clean up non-selected devices
                for (physicalDevice in devices) {
                    physicalDevice.cleanup()
                }
                if (selectedPhysicalDevice == null) {
                    throw RuntimeException("No suitable physical devices found")
                }
                logger.debug { "Selected device: [${selectedPhysicalDevice!!.deviceName}]" }
            }
            return selectedPhysicalDevice!!
        }

        protected fun getPhysicalDevices(stack: MemoryStack): PointerBuffer {
            val pPhysicalDevices: PointerBuffer
            // Get number of physical devices
            val intBuffer = stack.mallocInt(1)
            vkEnumeratePhysicalDevices(Instance.vkInstance, intBuffer, null)
                .vkCheck("Failed to get number of physical devices")

            val numDevices = intBuffer.get(0)
            logger.debug { "Detected {$numDevices} physical device(s)" }

            // Populate physical devices list pointer
            pPhysicalDevices = stack.mallocPointer(numDevices)
            vkEnumeratePhysicalDevices(Instance.vkInstance, intBuffer, pPhysicalDevices)
                .vkCheck("Failed to get physical devices")

            return pPhysicalDevices
        }
    }
}